File: code/ch11/11.3.1_characterReplacement.cpp
class Solution {
    public:
        int characterReplacement(string s, int k) {
            int res = 0, left = 0, right = 0, size = s.size(), localMaxFreq = 0;

            vector<int> count(26, 0);
            for (; right < size; right++) {
                count[s[right] - 'A'] += 1;
                localMaxFreq = max(localMaxFreq, count[s[right] - 'A']);

                if ((right - left + 1) - localMaxFreq > k) {
                    res = max(res, (right - left));
                    count[s[left] - 'A'] -= 1;
                    left += 1;
                    localMaxFreq = *(max_element(count.begin(), count.end()));
                }
            }

            return max(res, right - left);
        }
};

File: code/ch11/11.3.1_characterReplacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int res = 0, left = 0, right = 0;
        int[] counts = new int[26];

        while( right < s.length()) {
            counts[s.charAt(right) - 'A']++;
            int maxCharCount = getMostCharCount(counts);
            right++;
            if (right - left - maxCharCount > k) {
                counts[s.charAt(left) - 'A']--;
                left++;
            }
        }
        return right - left;
    }

    private int getMostCharCount(int[] counts) {
        int ans = 0;
        for (int n : counts)
            ans = Math.max(ans, n);
        return ans;
    }
}